Framework/Source/API/ConstantBuffer.cpp

```diff
@@ -86,250 +86,9 @@ namespace Falcor
-                    numMembers = pArrayType->getArraySize();
-                    memberSize = pArrayType->getArrayStride();
-
-                    if (elementBasicType)
-                    { 
-                        memberType = elementBasicType->getType();
-                    }
-                    else
-                    {
-                        // for special case of array of structures
-                        baseTypeIsStruct = true;
-                    }
-                }
-                else if (!pStructType)
-                {
-                    // Other types could be presented here
-                    return;
-                }
-            }
-            else
-            {
-                // information if only basic type
-                memberType = pBasicType->getType();
-                memberSize = pBasicType->getSize();
-            }
-
-            
-            // Display member of the array
-            int32_t memberIndex = 0;
-            std::string displayName = memberName;
-            
-            if (numMembers > 1)
-            {
-                // display information for specific index of array
-                memberIndex = mGuiArrayIndices[displayName];
-                pGui->addIntVar((std::string("Index (Size : ") + std::to_string(numMembers) + ") ").c_str(), memberIndex, 0, static_cast<int>(numMembers) - 1);
-                currentOffset += (memberSize * memberIndex);
-                displayName.append("[").append(std::to_string(memberIndex)).append("]");
-            }
-
-            if (baseTypeIsStruct)
-            {
-                // For arrays of structs, display dropdown for struct before recursing through struct members
-                if (pGui->beginGroup(displayName))
-                {
-                    displayName.push_back('.');
-                    renderUIInternal(pGui, pArrayType->getType()->asStructType(), displayName, currentOffset);
-                    pGui->endGroup();
-                }
-            }
-            else
-            {
-                // for basic types
-                renderUIMemberInternal(pGui, displayName, currentOffset, memberSize, to_string(memberType), memberType);
-            }
-            
-            currentOffset += memberSize;
-            
-
-            if (arrayGroupStatus)
-            {
-                pGui->endGroup();
-            }
-        }
-    }
-
-    void ConstantBuffer::renderUI(Gui* pGui)
-    {
-        const ReflectionStructType* pStruct = mpReflector->asResourceType()->getStructType()->asStructType();
-
-        if (pGui->beginGroup(std::string("ConstantBuffer: ").append(mName)))
-        {
-            pGui->addSeparator();
-
-            // begin recursion on first struct
-            renderUIInternal(pGui, pStruct, "", 0);
-
-            pGui->endGroup();
-        }
+        VariablesBufferUI variablesBufferUI(*this);
```

Comment:
What's the overhead of constructing a new object? If it's too high, maybe lazily allocate the UI object

OK, I took a look and the overhead is practically zero, so it's fine. Just be aware that in some instances the overhead can be high.